File: backend/services/protocol.py
"""CRV protocol state machine.

Pure functions over an ordered event history. No database access, so the
gating rules are testable in isolation and mirrored in the frontend
(frontend/lib/protocol.ts).

Rules follow the 1986 CRV manual [CRV-MANUAL], see docs/research/PROTOCOLS.md:
- Sessions start at Stage I and progress in order.
- Stage II opens only after a complete ideogram/A/B trio exists in Stage I.
- Each later stage opens only after the prior stage recorded substantive
  structure and an explicit stage_advance.
- After an AOL declaration, signal events are refused until the AOL break
  is objectified.
"""

from dataclasses import dataclass

from core.models.rv import EventKind, Protocol

CRV_STAGES = (1, 2, 3, 4, 5, 6)

# Event kinds that are legal in any stage of any protocol.
ALWAYS_ALLOWED: frozenset[EventKind] = frozenset(
    {
        EventKind.AOL,
        EventKind.AOL_BREAK,
        EventKind.BREAK,
        EventKind.MONITOR_PROMPT,
        EventKind.VIEWER_NOTE,
    }
)

# Signal-line event kinds a viewer produces. Refused while an AOL is open.
SIGNAL_KINDS: frozenset[EventKind] = frozenset(
    {
        EventKind.IDEOGRAM,
        EventKind.IDEOGRAM_A,
        EventKind.IDEOGRAM_B,
        EventKind.SENSORY,
        EventKind.DIMENSIONAL,
        EventKind.AESTHETIC_IMPACT,
        EventKind.EMOTIONAL_IMPACT,
        EventKind.TANGIBLE,
        EventKind.INTANGIBLE,
        EventKind.AOL_SIGNAL,
        EventKind.SKETCH,
    }
)

# Per-stage signal vocabulary for CRV [CRV-MANUAL].
CRV_STAGE_ALLOWED: dict[int, frozenset[EventKind]] = {
    1: frozenset({EventKind.IDEOGRAM, EventKind.IDEOGRAM_A, EventKind.IDEOGRAM_B}),
    2: frozenset({EventKind.SENSORY, EventKind.AESTHETIC_IMPACT, EventKind.SKETCH}),
    3: frozenset(
        {
            EventKind.DIMENSIONAL,
            EventKind.AESTHETIC_IMPACT,
            EventKind.SKETCH,
        }
    ),
    4: frozenset(
        {
            EventKind.SENSORY,
            EventKind.DIMENSIONAL,
            EventKind.AESTHETIC_IMPACT,
            EventKind.EMOTIONAL_IMPACT,
            EventKind.TANGIBLE,
            EventKind.INTANGIBLE,
            EventKind.AOL_SIGNAL,
            EventKind.SKETCH,
        }
    ),
    5: frozenset(
        {
            EventKind.SENSORY,
            EventKind.DIMENSIONAL,
            EventKind.TANGIBLE,
            EventKind.INTANGIBLE,
            EventKind.SKETCH,
        }
    ),
    6: frozenset(
        {
            EventKind.DIMENSIONAL,
            EventKind.TANGIBLE,
            EventKind.INTANGIBLE,
            EventKind.SKETCH,
        }
    ),
}

# Free-form protocols keep blindness, lock, and the typed transcript
# but skip CRV stage gating. Shared by ERV (narrative), ARV (future
# feedback photograph), and WRV (written / phonetic) [CIA-BRIEF].
FREEFORM_ALLOWED: frozenset[EventKind] = frozenset(
    {
        EventKind.SENSORY,
        EventKind.DIMENSIONAL,
        EventKind.EMOTIONAL_IMPACT,
        EventKind.SKETCH,
    }
)
ERV_ALLOWED = FREEFORM_ALLOWED
ARV_ALLOWED = FREEFORM_ALLOWED
WRV_ALLOWED = FREEFORM_ALLOWED
FREEFORM_PROTOCOLS: frozenset[Protocol] = frozenset(
    {Protocol.ERV, Protocol.ARV, Protocol.WRV}
)


@dataclass(frozen=True)
class EventView:
    """The slice of a transcript event the state machine needs."""

    kind: EventKind
    stage: int | None


@dataclass(frozen=True)
class Refusal:
    code: str
    message: str


def open_aol(events: list[EventView]) -> bool:
    """True when the latest AOL declaration has no AOL break after it."""
    for event in reversed(events):
        if event.kind == EventKind.AOL_BREAK:
            return False
        if event.kind == EventKind.AOL:
            return True
    return False


def stage_one_complete(events: list[EventView]) -> bool:
    """A complete ideogram trio exists: ideogram, then A, then B."""
    want = [EventKind.IDEOGRAM, EventKind.IDEOGRAM_A, EventKind.IDEOGRAM_B]
    idx = 0
    for event in events:
        if event.stage == 1 and event.kind == want[idx]:
            idx += 1
            if idx == len(want):
                return True
    return False


# What "substantive structure" means per stage, for advancement [CRV-MANUAL].
STAGE_ADVANCE_REQUIREMENT: dict[int, frozenset[EventKind]] = {
    2: frozenset({EventKind.SENSORY}),
    3: frozenset({EventKind.DIMENSIONAL, EventKind.SKETCH}),
    4: frozenset(
        {
            EventKind.SENSORY,
            EventKind.DIMENSIONAL,
            EventKind.AESTHETIC_IMPACT,
            EventKind.EMOTIONAL_IMPACT,
            EventKind.TANGIBLE,
            EventKind.INTANGIBLE,
            EventKind.AOL_SIGNAL,
        }
    ),
    5: frozenset({EventKind.TANGIBLE, EventKind.INTANGIBLE}),
}


def can_advance_stage(current_stage: int, events: list[EventView]) -> Refusal | None:
    """Whether the session may advance from current_stage to the next."""
    if current_stage >= 6:
        return Refusal("last_stage", "Stage VI is the final stage.")
    if open_aol(events):
        return Refusal(
            "aol_open",
            "Objectify the AOL break before advancing. Declared overlay must "
            "be set aside in structure.",
        )
    if current_stage == 1:
        if not stage_one_complete(events):
            return Refusal(
                "stage_one_incomplete",
                "Stage I is incomplete. Objectify the ideogram, then the A "
                "component, then the B component.",
            )
        return None
    required = STAGE_ADVANCE_REQUIREMENT[current_stage]
    for event in events:
        if event.stage == current_stage and event.kind in required:
            return None
    return Refusal(
        "stage_incomplete",
        f"Stage {current_stage} has no objectified structure yet. Record "
        "stage data before advancing.",
    )


def validate_event(
    protocol: Protocol,
    current_stage: int | None,
    kind: EventKind,
    events: list[EventView],
) -> Refusal | None:
    """Whether an event of this kind may be appended right now.

    Returns None when legal, or a Refusal explaining the structure rule.
    Lifecycle kinds (cue, stage_advance, lock, feedback_view) are engine
    internal and validated by their own code paths.
    """
    if kind in ALWAYS_ALLOWED:
        return None

    if kind in SIGNAL_KINDS and open_aol(events):
        return Refusal(
            "aol_open",
            "AOL declared. Objectify the break and set it aside before "
            "returning to the signal line.",
        )

    if protocol in FREEFORM_PROTOCOLS:
        if kind in FREEFORM_ALLOWED:
            return None
        return Refusal(
            "not_in_protocol",
            f"{kind.value} is not part of the {protocol.value.upper()} "
            "transcript vocabulary.",
        )

    if protocol != Protocol.CRV:
        return Refusal(
            "not_in_protocol",
            f"Protocol {protocol.value} does not run structured sessions.",
        )

    if current_stage not in CRV_STAGES:
        return Refusal("no_stage", "Session has no active stage.")

    allowed = CRV_STAGE_ALLOWED[current_stage]
    if kind in allowed:
        # Ideogram decoding must run in order within Stage I.
        if kind == EventKind.IDEOGRAM_A and not _has_stage_event(
            events, 1, EventKind.IDEOGRAM
        ):
            return Refusal(
                "structure_order",
                "Objectify the ideogram before decoding its A component.",
            )
        if kind == EventKind.IDEOGRAM_B and not _has_stage_event(
            events, 1, EventKind.IDEOGRAM_A
        ):
            return Refusal(
                "structure_order",
                "Objectify the A component before the B component.",
            )
        return None

    for stage, kinds in CRV_STAGE_ALLOWED.items():
        if kind in kinds:
            direction = "ahead of" if stage > current_stage else "behind"
            return Refusal(
                "out_of_structure",
                f"{kind.value} is Stage {_roman(stage)} content, {direction} "
                f"the current Stage {_roman(current_stage)}. Stay in "
                "structure.",
            )
    return Refusal("unknown_kind", f"{kind.value} is not viewer content.")


def _has_stage_event(events: list[EventView], stage: int, kind: EventKind) -> bool:
    return any(e.stage == stage and e.kind == kind for e in events)


_ROMAN = {1: "I", 2: "II", 3: "III", 4: "IV", 5: "V", 6: "VI"}


def _roman(stage: int) -> str:
    return _ROMAN.get(stage, str(stage))

File: frontend/lib/protocol.ts
/**
 * CRV protocol rules, mirrored from backend/services/protocol.py.
 *
 * The backend is authoritative. This mirror exists so the chamber can
 * gate its own controls without a round trip, and so protocol gating is
 * testable in the frontend suite.
 */

export type EventKind =
  | "cue"
  | "ideogram"
  | "ideogram_a"
  | "ideogram_b"
  | "sensory"
  | "dimensional"
  | "aesthetic_impact"
  | "emotional_impact"
  | "tangible"
  | "intangible"
  | "aol"
  | "aol_break"
  | "aol_signal"
  | "sketch"
  | "monitor_prompt"
  | "viewer_note"
  | "break"
  | "stage_advance"
  | "lock"
  | "feedback_view";

export type Protocol = "crv" | "erv" | "arv" | "wrv";

export interface EventView {
  kind: EventKind;
  stage: number | null;
}

export const ALWAYS_ALLOWED: ReadonlySet<EventKind> = new Set([
  "aol",
  "aol_break",
  "break",
  "monitor_prompt",
  "viewer_note",
]);

export const SIGNAL_KINDS: ReadonlySet<EventKind> = new Set([
  "ideogram",
  "ideogram_a",
  "ideogram_b",
  "sensory",
  "dimensional",
  "aesthetic_impact",
  "emotional_impact",
  "tangible",
  "intangible",
  "aol_signal",
  "sketch",
]);

export const CRV_STAGE_ALLOWED: Record<number, ReadonlySet<EventKind>> = {
  1: new Set(["ideogram", "ideogram_a", "ideogram_b"]),
  2: new Set(["sensory", "aesthetic_impact", "sketch"]),
  3: new Set(["dimensional", "aesthetic_impact", "sketch"]),
  4: new Set([
    "sensory",
    "dimensional",
    "aesthetic_impact",
    "emotional_impact",
    "tangible",
    "intangible",
    "aol_signal",
    "sketch",
  ]),
  5: new Set(["sensory", "dimensional", "tangible", "intangible", "sketch"]),
  6: new Set(["dimensional", "tangible", "intangible", "sketch"]),
};

export const FREEFORM_ALLOWED: ReadonlySet<EventKind> = new Set([
  "sensory",
  "dimensional",
  "emotional_impact",
  "sketch",
]);

export const ERV_ALLOWED = FREEFORM_ALLOWED;
export const FREEFORM_PROTOCOLS: ReadonlySet<Protocol> = new Set([
  "erv",
  "arv",
  "wrv",
]);

export const PROTOCOL_BRIEF: Record<Protocol, string> = {
  crv: "Coordinate Remote Viewing. Stage I through VI, in structure.",
  erv: "Extended Remote Viewing. Free-form narrative. Blindness and lock still hold.",
  arv: "Associative Remote Viewing. Describe the photograph you will see at feedback.",
  wrv: "Written Remote Viewing. Phonetic fragments and automatic writing on the page.",
};

const STAGE_ADVANCE_REQUIREMENT: Record<number, ReadonlySet<EventKind>> = {
  2: new Set(["sensory"]),
  3: new Set(["dimensional", "sketch"]),
  4: new Set([
    "sensory",
    "dimensional",
    "aesthetic_impact",
    "emotional_impact",
    "tangible",
    "intangible",
    "aol_signal",
  ]),
  5: new Set(["tangible", "intangible"]),
};

export function openAol(events: EventView[]): boolean {
  for (let i = events.length - 1; i >= 0; i--) {
    if (events[i].kind === "aol_break") return false;
    if (events[i].kind === "aol") return true;
  }
  return false;
}

export function stageOneComplete(events: EventView[]): boolean {
  const want: EventKind[] = ["ideogram", "ideogram_a", "ideogram_b"];
  let idx = 0;
  for (const event of events) {
    if (event.stage === 1 && event.kind === want[idx]) {
      idx += 1;
      if (idx === want.length) return true;
    }
  }
  return false;
}

export function canAdvanceStage(
  currentStage: number,
  events: EventView[],
): boolean {
  if (currentStage >= 6) return false;
  if (openAol(events)) return false;
  if (currentStage === 1) return stageOneComplete(events);
  const required = STAGE_ADVANCE_REQUIREMENT[currentStage];
  return events.some(
    (event) => event.stage === currentStage && required.has(event.kind),
  );
}

function hasStageEvent(
  events: EventView[],
  stage: number,
  kind: EventKind,
): boolean {
  return events.some((event) => event.stage === stage && event.kind === kind);
}

/** Whether an event of this kind may be recorded right now. */
export function canRecord(
  protocol: Protocol,
  currentStage: number | null,
  kind: EventKind,
  events: EventView[],
): boolean {
  if (ALWAYS_ALLOWED.has(kind)) return true;
  if (SIGNAL_KINDS.has(kind) && openAol(events)) return false;

  if (FREEFORM_PROTOCOLS.has(protocol)) return FREEFORM_ALLOWED.has(kind);
  if (protocol !== "crv") return false;

  if (currentStage === null || !(currentStage in CRV_STAGE_ALLOWED)) {
    return false;
  }
  if (!CRV_STAGE_ALLOWED[currentStage].has(kind)) return false;

  if (kind === "ideogram_a" && !hasStageEvent(events, 1, "ideogram")) {
    return false;
  }
  if (kind === "ideogram_b" && !hasStageEvent(events, 1, "ideogram_a")) {
    return false;
  }
  return true;
}

export const STAGE_ROMAN: Record<number, string> = {
  1: "I",
  2: "II",
  3: "III",
  4: "IV",
  5: "V",
  6: "VI",
};

export const STAGE_TITLES: Record<number, string> = {
  1: "Ideogram",
  2: "Sensory",
  3: "Dimension",
  4: "Matrix",
  5: "Interrogation",
  6: "Rendering",
};

export const KIND_LABELS: Partial<Record<EventKind, string>> = {
  ideogram: "Ideogram",
  ideogram_a: "A component",
  ideogram_b: "B component",
  sensory: "Sensory",
  dimensional: "Dimensional",
  aesthetic_impact: "Aesthetic impact",
  emotional_impact: "Emotional impact",
  tangible: "Tangible",
  intangible: "Intangible",
  aol_signal: "AOL / signal",
  sketch: "Sketch",
  aol: "AOL",
  aol_break: "AOL break",
  viewer_note: "Note",
  break: "Break",
  monitor_prompt: "Monitor",
  cue: "Cue",
  stage_advance: "Stage advance",
  lock: "Lock",
  feedback_view: "Feedback",
};

export const LEADING_MARKERS = [
  "looks like",
  "is it a",
  "could it be",
  "maybe it's",
  "maybe it is",
  "i think it",
  "the target",
  "what do you think it is",
  "what is it",
] as const;

export function isLeadingPatter(text: string): boolean {
  const lowered = text.toLowerCase();
  return LEADING_MARKERS.some((marker) => lowered.includes(marker));
}

export const HUMAN_PATTER = [
  "Take the cue.",
  "Objectify.",
  "Stay low level. Single words.",
  "AOL break. Declare it, set it aside.",
  "Sketch what presents itself.",
  "I will keep the record.",
  "Back to the signal line when ready.",
] as const;
